pytorch_lightning/trainer/connectors/logger_connector.py

```diff
@@ -171,14 +171,20 @@ def _log_on_evaluation_epoch_end_metrics(self, epoch_logs):
         # --------------------------------
         # each dataloader aggregated metrics
         # now we log all of them
+
+        device = None
+
+        if self.trainer.distributed_backend == 'dp':
+            device = self.trainer.root_gpu
+
         for dl_idx, dl_metrics in enumerate(step_metrics):
             if len(dl_metrics) == 0:
                 # Ensure custom logged metrics are included if not included with step metrics
                 if len(epoch_logger_metrics) > 0:
                     self.eval_loop_results.append(epoch_logger_metrics)
                 continue
 
-            reduced_epoch_metrics = dl_metrics[0].__class__.reduce_on_epoch_end(dl_metrics)
+            reduced_epoch_metrics = dl_metrics[0].__class__.reduce_on_epoch_end(dl_metrics, device=device)
```

Comment:
This fix really helps me. 

Besides I encounter this problem in `log_train_epoch_end_metrics`, it also needs this fix.

> https://github.com/PyTorchLightning/pytorch-lightning/blob/360b3d88440d455e2235da33616014f4f45d8bd0/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L420

Reply:
Is it the same bug? Because this one did not occur for me. I'll dig into this!

Reply:
This is my crash outputs:
```bash
Traceback (most recent call last):
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 485, in train
    self.train_loop.run_training_epoch()
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/training_loop.py", line 610, in run_training_epoch
    epoch_output, self.checkpoint_accumulator, self.early_stopping_accumulator, self.num_optimizers
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/logger_connector.py", line 352, in log_train_epoch_end_metrics
    epoch_log_metrics, epoch_progress_bar_metrics = self.__auto_reduce_results_on_epoch_end(epoch_output)
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/logger_connector.py", line 470, in __auto_reduce_results_on_epoch_end
    opt_outputs = time_reduced_outputs[0].__class__.reduce_on_epoch_end(time_reduced_outputs, device=device)
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/core/step_result.py", line 463, in reduce_on_epoch_end
python-BaseException
    recursive_stack(result, device=device)
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/core/step_result.py", line 602, in recursive_stack
    result[k] = collate_tensors(v, device=device)
  File "/data2/zhaoxiaoming/anaconda3/lib/python3.7/site-packages/pytorch_lightning/core/step_result.py", line 628, in collate_tensors
    return torch.stack(items)
RuntimeError: All input tensors must be on the same device. Received cuda:2 and cuda:3
```

Reply:
Hello, when I modified `reduced_epoch_metrics` = `dl_metrics[0].__class__.reduce_on_epoch_end(dl_metrics)` to
            `reduced_epoch_metrics` = dl_metrics[0].__class__.reduce_on_epoch_end(dl_metrics, device=device)`, there is a new error raised that **NameError: name 'device' is not defined**. Do you know what is happened?